In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df_org = pd.read_csv('/content/sample_data/mnist_train_small.csv')
print(df_org.shape)
df_org.head()

(19999, 785)


6  0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  ...  0.581  0.582  0.583  \
0  5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
1  7  0    0    0    0    0    0    0    0    0  ...      0      0      0   
2  9  0    0    0    0    0    0    0    0    0  ...      0      0      0   
3  5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
4  2  0    0    0    0    0    0    0    0    0  ...      0      0      0   

   0.584  0.585  0.586  0.587  0.588  0.589  0.590  
0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0  

[5 rows x 785 columns]

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
X, y = np.arange(10).reshape((5, 2)), range(5)
X
list(y)

[0, 1, 2, 3, 4]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42)
X_train
y_train
X_test
y_test

[1, 4]

In [ ]:
train_test_split(y, shuffle=False)

[[0, 1, 2], [3, 4]]

In [ ]:
from sklearn import datasets
iris = datasets.load_iris(as_frame=True)
X, y = iris['data'], iris['target']
X.head()
y.head()

0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size=0.33, random_state=42)
X_train.head()
y_train.head()
X_test.head()
y_test.head()

73     1
18     0
118    2
78     1
76     1
Name: target, dtype: int64

# Task
Split the dataset into training and test sets using Scikit-Learn, conduct data exploration, data preprocessing, and feature engineering if necessary, train and test a decision tree model using Scikit-Learn, and check the documentation to identify the most important hyperparameters, attributes, and methods of the model and use them in practice.

## Split the dataset

### Subtask:
Split the `df_org` DataFrame into training and test sets.


**Reasoning**:
Separate the target variable from the features and split the data into training and testing sets using an 80/20 split.



In [ ]:
X = df_org.drop(columns=[df_org.columns[0]])
y = df_org[df_org.columns[0]]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Data exploration

### Subtask:
Explore the training data to understand its structure, distributions, and potential issues.


**Reasoning**:
Display the first few rows, get a concise summary, generate descriptive statistics of the training data, and check the distribution of the target variable.



In [ ]:
display(X_train.head())
display(X_train.info())
display(X_train.describe())
display(y_train.value_counts())

0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  ...  0.581  0.582  \
5894  0    0    0    0    0    0    0    0    0    0  ...      0      0   
3728  0    0    0    0    0    0    0    0    0    0  ...      0      0   
8958  0    0    0    0    0    0    0    0    0    0  ...      0      0   
7671  0    0    0    0    0    0    0    0    0    0  ...      0      0   
5999  0    0    0    0    0    0    0    0    0    0  ...      0      0   

      0.583  0.584  0.585  0.586  0.587  0.588  0.589  0.590  
5894      0      0      0      0      0      0      0      0  
3728      0      0      0      0      0      0      0      0  
8958      0      0      0      0      0      0      0      0  
7671      0      0      0      0      0      0      0      0  
5999      0      0      0      0      0      0      0      0  

[5 rows x 784 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 15999 entries, 5894 to 15795
Columns: 784 entries, 0 to 0.590
dtypes: int64(784)
memory usage: 95.8 MB


None

0      0.1      0.2      0.3      0.4      0.5      0.6      0.7  \
count  15999.0  15999.0  15999.0  15999.0  15999.0  15999.0  15999.0  15999.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           0.8      0.9  ...         0.581         0.582         0.583  \
count  15999.0  15999.0  ...  15999.000000  15999.000000  15999.000000   
mean       0.0      0.0  ...      0.226952      0.097506      0.029814   
std        0.0      0.0  ...      6.638906      4.250588      2.279736   
min        0.0      0.0  ...      0.000000      0.000000      0.000000   
25%        0.0      0.0  ...      0.000000      0.000000      0.000000   
50%        0.0      0.0  ...      0.000000      0.000000      0.000000   
75%        0.0      0.0  ...      0.000000      0.000000      0.000000   
max        0.0      0.0  ...    253.000000    254.000000    199.000000   

              0.584         0.585         0.586    0.587    0.588    0.589  \
count  15999.000000  15999.000000  15999.000000  15999.0  15999.0  15999.0   
mean       0.010376      0.025502      0.003625      0.0      0.0      0.0   
std        0.870983      2.348306      0.458545      0.0      0.0      0.0   
min        0.000000      0.000000      0.000000      0.0      0.0      0.0   
25%        0.000000      0.000000      0.000000      0.0      0.0      0.0   
50%        0.000000      0.000000      0.000000      0.0      0.0      0.0   
75%        0.000000      0.000000      0.000000      0.0      0.0      0.0   
max       79.000000    254.000000     58.000000      0.0      0.0      0.0   

         0.590  
count  15999.0  
mean       0.0  
std        0.0  
min        0.0  
25%        0.0  
50%        0.0  
75%        0.0  
max        0.0  

[8 rows x 784 columns]

6
1    1807
7    1688
6    1626
3    1615
9    1608
2    1599
0    1585
8    1528
4    1523
5    1420
Name: count, dtype: int64

## Data preprocessing

### Subtask:
Preprocess the data as needed (e.g., handle missing values, scale features).


**Reasoning**:
Scale the pixel intensity values in X_train and X_test to the range of 0 to 1 by dividing by 255.



In [ ]:
X_train = X_train / 255.0
X_test = X_test / 255.0
display(X_train.head())
display(X_test.head())

0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  ...  0.581  0.582  \
5894  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
3728  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
8958  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
7671  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
5999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   

      0.583  0.584  0.585  0.586  0.587  0.588  0.589  0.590  
5894    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
3728    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
8958    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
7671    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
5999    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 784 columns]

0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  ...  0.581  0.582  \
10650  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
2041   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
8668   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
1114   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
13902  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   

       0.583  0.584  0.585  0.586  0.587  0.588  0.589  0.590  
10650    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
2041     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
8668     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
1114     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
13902    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 784 columns]

## Feature engineering

### Subtask:
Create new features from existing ones if beneficial for the model.


## Train a decision tree model

### Subtask:
Train a Decision Tree classifier on the training data.


**Reasoning**:
Import the DecisionTreeClassifier and train the model on the training data.



In [ ]:
from sklearn.tree import DecisionTreeClassifier

decision_tree_model = DecisionTreeClassifier(random_state=42)
decision_tree_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

## Tune hyperparameters

### Subtask:
Identify and tune important hyperparameters of the Decision Tree model.


**Reasoning**:
Import necessary modules for hyperparameter tuning and define the parameter grid for the Decision Tree model.



In [ ]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'criterion': ['gini', 'entropy']
}

**Reasoning**:
Instantiate GridSearchCV with the Decision Tree model and the parameter grid, then fit it to the training data to find the best hyperparameters.



In [ ]:
grid_search = GridSearchCV(decision_tree_model, param_grid, cv=3, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=DecisionTreeClassifier(random_state=42), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 10, 20, 30],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10]},
             scoring='accuracy')

**Reasoning**:
Store the best model found by the tuning process and display the best parameters and the best score.



In [ ]:
best_decision_tree_model = grid_search.best_estimator_
print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation accuracy: ", grid_search.best_score_)

Best parameters found:  {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2}
Best cross-validation accuracy:  0.826426651665729


## Evaluate the model

### Subtask:
Evaluate the trained model on the test set using relevant metrics.


**Reasoning**:
Use the best trained model to make predictions on the test set and evaluate the model using accuracy, precision, recall, and F1-score.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred = best_decision_tree_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.8325
Precision: 0.8342
Recall: 0.8325
F1-score: 0.8325


## Inspect attributes and methods

### Subtask:
Explore important attributes and methods of the trained model.


**Reasoning**:
Explore the important attributes and methods of the trained decision tree model as requested in the instructions.



In [ ]:
# 1. Display classes_ attribute
print("Classes:", best_decision_tree_model.classes_)

# 2. Display feature_importances_ attribute
print("\nFeature Importances (first 20):", best_decision_tree_model.feature_importances_[:20])

# 3. Explore get_params() method
print("\nModel Parameters:", best_decision_tree_model.get_params())

# 4. Explore predict_proba() method on a small subset
X_test_subset = X_test.head(5)
predicted_probabilities = best_decision_tree_model.predict_proba(X_test_subset)
print("\nPredicted Probabilities for first 5 test samples:\n", predicted_probabilities)

Classes: [0 1 2 3 4 5 6 7 8 9]

Feature Importances (first 20): [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

Model Parameters: {'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'entropy', 'max_depth': None, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 2, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'monotonic_cst': None, 'random_state': 42, 'splitter': 'best'}

Predicted Probabilities for first 5 test samples:
 [[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]]


## Summary:

### Data Analysis Key Findings

*   The dataset was successfully split into training (80%) and test (20%) sets.
*   The training data consists of 15999 samples with 784 features (pixel data) and a target variable representing digits 0-9.
*   There are no missing values in the training data.
*   The class distribution in the training set is relatively balanced.
*   The pixel intensity values were successfully scaled from the range 0-255 to 0-1.
*   No additional feature engineering was performed as it was deemed unnecessary for this task.
*   A Decision Tree classifier was trained on the scaled training data.
*   Hyperparameter tuning using GridSearchCV identified the best parameters as `{'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2}` with a best cross-validation accuracy of approximately 0.8264.
*   The best trained model achieved an accuracy of 0.8325, precision of 0.8342, recall of 0.8325, and F1-score of 0.8325 on the test set.
*   Inspection of the model revealed the learned classes (0-9), feature importances (though the first 20 were zero), current parameters, and the ability to predict class probabilities.

### Insights or Next Steps

*   While the initial features showed zero importance, further investigation into the `feature_importances_` for other features could provide insights into which pixels are most discriminative for digit classification.
*   The model's performance is reasonable for a base Decision Tree. Exploring more complex models like Random Forests or Gradient Boosting, or using techniques specifically designed for image data (e.g., CNNs), could potentially improve accuracy.
